DIPs/DIP1005.md

```diff
@@ -0,0 +1,149 @@
+
+We propose an addition to the D language that allows the use of the keyword `import` as a pseudo-package name at the top level. When that syntax is used, it instructs the compiler to look up the symbol in the package/module that follows. To clarify by means of examples, the previous examples can be rewritten as:
+
+    void process(import.std.stdio.File input);
+    struct Buffered(Range) if (import.std.range.isInputRange!Range)
+    {
+        ...
+    }
+
+With this syntax, the `import` is effected only if the declared name is actually looked up, and only looks up the imported package/module for that particular symbol only. Of course, simple caching will make several imports of the same module as expensive as the first.
+
+The feature can be applied to any symbol that can be looked up in another module by means of existing `import` declarations. The usual lookup rules apply, for example either `import.std.range.isInputRange` or the more precise `import.std.range.primitives.isInputRange` may be used.
+
+#### Alternatives Considered
+
+There are a number of alternative approaches that have been (and some still are) considered.
+
+* Full lazy `import`s. Assume all `import`s are lazy without any change in the language. That would allow an idiom by which libraries use fully qualified names everywhere, and the `import`s would be effected only when names are looked up. This would allow scalability but not the dependency-carrying aspect.
+* Add syntax to allow for an optional `import` directive inside declarations:
+
+        void process(import std.stdio)(File input);
+        struct Buffered(import std.range)(Range) if (isInputRange!Range)
+        {
+            ...
+        }
+
+This has the advantage of being less verbose in case the same module is looked up several times. The disadvantages are a heavier and more ambiguous syntax (two sets of parens for nontemplates, three for templates) and an unclear relationship between the `import`ed entities and the symbols used in the declaration.
+* Use `import` with a syntax similar to `if` in the declaration:
+
+        void process(File input) import (std.stdio);
+        struct Buffered(import std.range)(Range) if (isInputRange!Range)
+        import (std.stdio)
+        {
+            ...
+        }
+Such an option is viable and has again the advantage of reducing repetition in case the same `import` is used multiple times. In addition, the `import` may be made available inside the definition of the function/data structure, when present.
+
+### Breaking changes / deprecation process
+
+We do not anticipate any breaking changes brought by this language addition.
```

Comment:
Let's try to "prove" this a little: because `import` is already a reserved keyword, symbol names cannot start with `import` and thus `import.modulename.Type` cannot occur in any code before this DIP.